core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -327,32 +373,41 @@ public boolean entryExpiredInMemoryFromIteration(InternalCacheEntry<K, V> entry,
    }
 
    @Override
-   public void handleInStoreExpiration(K key) {
-      if (expiring.putIfAbsent(key, key) == null) {
-         // Unfortunately stores don't pull the entry so we can't tell exactly why it expired and thus we have to remove
-         // the entire value.  Unfortunately this could cause a concurrent write to be undone
-         try {
-            cache.withFlags(Flag.SKIP_SHARED_CACHE_STORE).removeLifespanExpired(key, null, null).join();
-         } finally {
-            expiring.remove(key, key);
-         }
-      }
+   public CompletionStage<Void> handleInStoreExpirationInternal(K key) {
+      return handleInStoreExpirationInternal(key, null);
    }
 
    @Override
-   public void handleInStoreExpiration(MarshallableEntry<K, V> marshalledEntry) {
-      K key = marshalledEntry.getKey();
-      if (expiring.putIfAbsent(key, key) == null) {
-         try {
-            Metadata metadata = marshalledEntry.getMetadata();
-            cache.withFlags(Flag.SKIP_SHARED_CACHE_STORE)
-                 .removeLifespanExpired(key, marshalledEntry.getValue(),
-                                        metadata.lifespan() == -1 ? null : metadata.lifespan())
-                 .join();
-         } finally {
-            expiring.remove(key, key);
+   public CompletionStage<Void> handleInStoreExpirationInternal(MarshallableEntry<K, V> marshalledEntry) {
+      return handleInStoreExpirationInternal(marshalledEntry.getKey(), marshalledEntry);
+   }
+
+   private CompletionStage<Void> handleInStoreExpirationInternal(K key, MarshallableEntry<K, V> marshallableEntry) {
+      CompletableFuture<Boolean> completableFuture = new CompletableFuture<>();
+      CompletableFuture<Boolean> previousFuture = expiring.putIfAbsent(key, completableFuture);
+      if (previousFuture == null) {
+         completableFuture.whenComplete((ignore, t) -> expiring.remove(key));
```

Comment:
`expiring.remove(key)` can be moved to line 403 and avoid creating a new lambda :)

Reply:
Fixed.